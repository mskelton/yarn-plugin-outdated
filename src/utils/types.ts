export interface DependencyInfo {
  name: string
  current: string
  latest: string
  type: "dependencies" | "devDependencies"
}
