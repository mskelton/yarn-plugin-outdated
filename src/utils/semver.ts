import * as semver from "semver"

export const parseVersion = (version: string) =>
  semver.valid(semver.coerce(version))!
