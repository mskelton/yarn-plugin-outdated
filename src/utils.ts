import { Manifest } from "@yarnpkg/core"
import { fromUrl } from "hosted-git-info"
import semver from "semver"

export const truthy = Boolean as unknown as <T>(
  arg: T | undefined | null | false
) => arg is T

export function getHomepageURL({ raw: manifest }: Manifest): string | null {
  const repo = manifest.repository
  const repoURL = manifest.homepage
    ? manifest.homepage
    : typeof repo === "string"
    ? repo
    : typeof repo === "object" && typeof repo.url === "string"
    ? repo.url
    : null

  const info = repoURL ? fromUrl(repoURL) : undefined
  const commitish = info?.committish ? `#${info.committish}` : ""

  return info
    ? `https://${info.domain}/${info.user}/${info.project}${commitish}`
    : repoURL
}

/**
 * Because some packages have a pre-release version as their `latest` version,
 * we need to first check if the latest version is a pre-release. If it is,
 * we compare the current and latest directly, otherwise we coerce the current
 * version to remove any pre-release identifiers to determine if it is outdated.
 */
export function isVersionOutdated(current: string, latest: string) {
  return semver.parse(latest)!.prerelease.length
    ? semver.lt(current, latest)
    : semver.lt(semver.coerce(current)!, latest)
}
