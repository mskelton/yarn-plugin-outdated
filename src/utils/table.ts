import { Configuration, formatUtils } from "@yarnpkg/core"
import * as semver from "semver"
import { DependencyInfo } from "./types"

const columns = ["name", "current", "latest", "type"] as const

export type TableColumn = typeof columns[number]

export const headers: Record<TableColumn, string> = {
  current: "Current",
  latest: "Latest",
  name: "Package",
  type: "Package Type",
}

// Iterates through the dependencies to determine the necessary size
// of each column.
function getColumnSizes(
  dependencies: DependencyInfo[]
): Record<TableColumn, number> {
  const sizes = {
    current: headers.current.length,
    latest: headers.latest.length,
    name: headers.name.length,
    type: headers.type.length,
  }

  for (const dependency of dependencies) {
    for (const [key, value] of Object.entries(dependency)) {
      const acc = sizes[key as TableColumn]
      const cur = value.length

      sizes[key as TableColumn] = acc > cur ? acc : cur
    }
  }

  return sizes
}

function printHeader(
  configuration: Configuration,
  sizes: Record<TableColumn, number>
) {
  const format = (key: TableColumn) =>
    formatUtils.pretty(configuration, headers[key].padEnd(sizes[key]), "bold")

  printRow({
    current: format("current"),
    latest: format("latest"),
    name: format("name"),
    type: format("type"),
  })
}

function printRow(row: Record<TableColumn, string>) {
  console.log(columns.map((column) => row[column]).join("   "))
}

export function printTable(
  configuration: Configuration,
  dependencies: DependencyInfo[]
) {
  const sizes = getColumnSizes(dependencies)
  printHeader(configuration, sizes)

  dependencies.forEach((dep) => {
    const current = semver.parse(dep.current)!
    const latest = semver.parse(dep.latest)!

    const color =
      latest.major > current.major
        ? "red"
        : latest.minor > current.minor
        ? "yellow"
        : "green"

    printRow({
      current: dep.current.padEnd(sizes.current),
      latest: dep.latest.padEnd(sizes.latest),
      name: formatUtils.pretty(
        configuration,
        dep.name.padEnd(sizes.name),
        color
      ),
      type: dep.type.padEnd(sizes.type),
    })
  })
}
