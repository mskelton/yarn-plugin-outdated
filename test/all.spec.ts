import { makeTemporaryEnv } from "./utils/env"

const candidates = {
  "package-a": "1.1.0",
  "package-b": "1.3.0",
  "package-c": "1.2.1",
  "package-d": "2.0.0",
}

describe.only("yarn outdated --all", () => {
  it("should include packages from all workspaces", async () => {
    const output = await run("outdated --all", {
      candidates,
      manifests: [
        {
          dependencies: {
            "package-a": "1.1.0",
            "package-b": "^1.1.1",
            "package-c": "~1.2.0",
          },
          name: {
            name: "workspace-a",
            scope: null,
          },
        },
        {
          devDependencies: {
            "package-c": "~1.2.0",
            "package-d": "> 1.1.0",
          },
          name: {
            name: "workspace-b",
            scope: "my-scope",
          },
        },
      ],
    })

    expect(output).toMatchSnapshot()
  })

  it.only("should fallback to computed workspace name", async () => {
    const { registry, run, writeFile, writeJson } = await makeTemporaryEnv()

    const packageJSON = {
      dependencies: {
        a: "1.1.0",
        b: "^1.1.1",
      },
    }

    const lockfile = `
__metadata:
  version: 4
  cacheKey: 8

"a@npm:1.1.0":
  version: 1.1.0
  resolution: "a@npm:1.1.0"
  checksum: b66e05fa25e42172a4f32aa8a5372203227f1b0e1c5d7879f45217a0fd0a5506a0f53f677ee532b8c92e53ce8404fdd60ba4cd83cfb806a19ad6472e6b75170b
  languageName: node
  linkType: hard

"b@npm:^1.1.1":
  version: 1.1.1
  resolution: "b@npm:^1.1.1"
  checksum: b66e05fa25e42172a4f32aa8a5372203227f1b0e1c5d7879f45217a0fd0a5506a0f53f677ee532b8c92e53ce8404fdd60ba4cd83cfb806a19ad6472e6b75170b
  languageName: node
  linkType: hard
`

    registry.packages = { a: "1.1.0", b: "1.3.0" }
    await writeJson("package.json", packageJSON)
    await writeFile("yarn.lock", lockfile)
    await run("install")

    const { stderr, stdout } = await run("outdated -a")
    expect(stdout).toMatchSnapshot()
    expect(stderr).toBeNull()
  })
})
