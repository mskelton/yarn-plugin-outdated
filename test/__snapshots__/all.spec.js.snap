// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yarn outdated --all should fallback to computed workspace name 1`] = `
"Checking for outdated dependencies
--- PROGRESS ---
--- SEPARATOR ---
INFO: Package     Current   Latest   Workspace       Package Type
INFO: package-b   1.1.1     1.3.0    computed-name   dependencies
--- SEPARATOR ---
WARN: 1 dependency is out of date
"
`;

exports[`yarn outdated --all should include packages from all workspaces 1`] = `
"Checking for outdated dependencies
--- PROGRESS ---
--- SEPARATOR ---
INFO: Package     Current   Latest   Workspace               Package Type   
INFO: package-b   1.1.1     1.3.0    workspace-a             dependencies   
INFO: package-c   1.2.0     1.2.1    workspace-a             dependencies   
INFO: package-c   1.2.0     1.2.1    @my-scope/workspace-b   devDependencies
INFO: package-d   1.1.0     2.0.0    @my-scope/workspace-b   devDependencies
--- SEPARATOR ---
WARN: 4 dependencies are out of date
"
`;
